src/main/java/com/aws/greengrass/shadowmanager/sync/strategy/RealTimeSyncStrategy.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.shadowmanager.sync.strategy;
+
+import com.aws.greengrass.logging.api.Logger;
+import com.aws.greengrass.logging.impl.LogManager;
+import com.aws.greengrass.shadowmanager.exception.RetryableException;
+import com.aws.greengrass.shadowmanager.exception.UnknownShadowException;
+import com.aws.greengrass.shadowmanager.sync.Retryer;
+import com.aws.greengrass.shadowmanager.sync.model.FullShadowSyncRequest;
+import com.aws.greengrass.shadowmanager.sync.model.SyncContext;
+import com.aws.greengrass.shadowmanager.sync.model.SyncRequest;
+import com.aws.greengrass.util.RetryUtils;
+import software.amazon.awssdk.aws.greengrass.model.ConflictError;
+import software.amazon.awssdk.services.iotdataplane.model.ConflictException;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.aws.greengrass.shadowmanager.model.Constants.LOG_SHADOW_NAME_KEY;
+import static com.aws.greengrass.shadowmanager.model.Constants.LOG_THING_NAME_KEY;
+
+public class RealTimeSyncStrategy extends BaseSyncStrategy implements SyncStrategy {
+
+    private static final Logger logger = LogManager.getLogger(RealTimeSyncStrategy.class);
+    private final ExecutorService syncExecutorService;
+
+    /**
+     * Indicates whether syncing is running or not.
+     */
+    AtomicBoolean syncing = new AtomicBoolean(false);
+
+
+    /**
+     * Constructor.
+     *
+     * @param executorService executor service.
+     * @param retryer         The retryer object.
+     */
+    public RealTimeSyncStrategy(ExecutorService executorService, Retryer retryer) {
+        super(retryer);
+        this.syncExecutorService = executorService;
+    }
+
+    /**
+     * Constructor for testing.
+     *
+     * @param executorService executor service.
+     * @param retryer         The retryer object.
+     * @param retryConfig     The retryer configuration.
+     */
+    public RealTimeSyncStrategy(ExecutorService executorService, Retryer retryer,
+                                RetryUtils.RetryConfig retryConfig) {
+        super(retryer, retryConfig);
+        this.syncExecutorService = executorService;
+    }
+
+    /**
+     * Starts syncing the shadows based on the strategy.
+     *
+     * @param context         an context object for syncing
+     * @param syncParallelism number of threads to use for syncing
+     */
+    @Override
+    public void startSync(SyncContext context, int syncParallelism) {
+        synchronized (lifecycleLock) {
+            if (syncing.compareAndSet(false, true)) {
```

Comment:
nit: could this be moved to the Base class (especially if it is made abstract) and then you have a `doStop` and `doStart` that don't need to do this synchronization and compare and set?

e.g.: 
```
public void startSync(...) {
        synchronized (lifecycleLock) {
            if (syncing.compareAndSet(false, true)) {
              doStart(context, syncParallelism);
            } else {
               logger.atDebug(SYNC_EVENT_TYPE).log("Already started syncing");
            }
    }
}

abstract void doStart();
```